File: ruby/dec_05/dec_05.rb
# Advent of Code 2025 Day 5 - Cafeteria Inventory (Ruby)

sections = File.read("input_05.txt").strip.split("\n\n")

# Parse ranges
fresh_ranges = sections[0].lines.map do |line|
  start, end_ = line.strip.split("-").map(&:to_i)
  [start, end_]
end

# Parse IDs
available_ids = sections[1].lines.map(&:strip).reject(&:empty?).map(&:to_i)

# Part I: Count fresh available IDs
result1 = available_ids.count do |id|
  fresh_ranges.any? { |start, end_| start <= id && id <= end_ }
end

# Part II: ids in any range
merged = fresh_ranges.sort_by(&:first)
merged_ranges = []
merged.each do |start, end_|
  if merged_ranges.empty? || start > merged_ranges[-1][1] + 1
    merged_ranges << [start, end_]
  else
    merged_ranges[-1][1] = [merged_ranges[-1][1], end_].max
  end
end
result2 = merged_ranges.sum { |start, end_| end_ - start + 1 }

puts "First: #{result1}"
puts "Second: #{result2}"

File: python/dec_05/dec_05.py
# Advent of Code 2025 Day 5 - Cafeteria Inventory

with open("input_05.txt", "r") as f:
    sections = f.read().strip().split("\n\n")

# Parse ranges
fresh_ranges = []
for line in sections[0].splitlines():
    start, end = map(int, line.split("-"))
    fresh_ranges.append((start, end))

# Parse IDs
available_ids = [int(line) for line in sections[1].splitlines() if line.strip()]

# Part I: Count fresh available IDs
result1 = 0
for id_ in available_ids:
    if any(start <= id_ <= end for start, end in fresh_ranges):
        result1 += 1


# Part II: ids in any range
def part_two(intervals):
    intervals = sorted(intervals)
    merged = []
    for start, end in intervals:
        if not merged or start > merged[-1][1] + 1:
            merged.append([start, end])
        else:
            merged[-1][1] = max(merged[-1][1], end)
    return merged

merged_intervals = part_two(fresh_ranges)
result2 = sum(end - start + 1 for start, end in merged_intervals)

print(f"First: {result1}")
print(f"Second: {result2}")
